pinot-controller/src/main/java/org/apache/pinot/controller/cursors/ResponseStoreCleaner.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.helix.core.periodictask.ControllerPeriodicTask;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ResponseStoreCleaner extends ControllerPeriodicTask<Void> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponseStoreCleaner.class);
+  private static final int TIMEOUT_MS = 3000;
+  private static final String QUERY_RESULT_STORE = "%s://%s:%d/responseStore";
+  private static final String DELETE_QUERY_RESULT = "%s://%s:%d/responseStore/%s";
+  public static final String CLEAN_AT_TIME = "response.store.cleaner.clean.at.ms";
+  private final ControllerConf _controllerConf;
+  private final Executor _executor;
+  private final PoolingHttpClientConnectionManager _connectionManager;
+
+  public ResponseStoreCleaner(ControllerConf config, PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerMetrics controllerMetrics, Executor executor,
+      PoolingHttpClientConnectionManager connectionManager) {
+    super("ResultStoreCleaner", getFrequencyInSeconds(config), getInitialDelayInSeconds(config),
+        pinotHelixResourceManager, leadControllerManager, controllerMetrics);
+    _controllerConf = config;
+    _executor = executor;
+    _connectionManager = connectionManager;
+  }
+
+  private static long getInitialDelayInSeconds(ControllerConf config) {
+    long initialDelay = config.getPeriodicTaskInitialDelayInSeconds();
+    String resultStoreCleanerTaskInitialDelay =
+        config.getProperty(CommonConstants.CursorConfigs.RESULT_STORE_CLEANER_INITIAL_DELAY);
+    if (resultStoreCleanerTaskInitialDelay != null) {
+      initialDelay = TimeUnit.SECONDS.convert(TimeUtils.convertPeriodToMillis(resultStoreCleanerTaskInitialDelay),
+          TimeUnit.MILLISECONDS);
+    }
+    return initialDelay;
+  }
+
+  private static long getFrequencyInSeconds(ControllerConf config) {
+    long frequencyInSeconds = 0;
+    String resultStoreCleanerTaskPeriod =
+        config.getProperty(CommonConstants.CursorConfigs.RESULT_STORE_CLEANER_FREQUENCY_PERIOD);
+    if (resultStoreCleanerTaskPeriod != null) {
+      frequencyInSeconds = TimeUnit.SECONDS.convert(TimeUtils.convertPeriodToMillis(resultStoreCleanerTaskPeriod),
+          TimeUnit.MILLISECONDS);
+    }
+
+    return frequencyInSeconds;
+  }
+
+  @Override
+  protected void processTables(List<String> tableNamesWithType, Properties periodicTaskProperties) {
+    long cleanAtMs = System.currentTimeMillis();
+    String cleanAtMsStr = periodicTaskProperties.getProperty(CLEAN_AT_TIME);
+    if (cleanAtMsStr != null) {
+      cleanAtMs = Long.parseLong(cleanAtMsStr);
+    }
+    doClean(cleanAtMs);
+  }
+
+  public void doClean(long currentTime) {
```

Comment:
nit: this method declares `brokerResponses` and `responses` variables. The first is the list of cursors open on each broker, which the method iterates to then ask each broker to delete it if needed. `responses` store the result of that second request to brokers.

It may be subjective, but the names are not very clear. Both are responses from browsers. I would appreciate to look for better names, like `knownCursors` and `cleanResponses` or something like that.

Reply:
Done